File: src/count_discordant_reads.py
#!/usr/bin/env python3

import argparse
import os

import pandas as pd

from pipeline.tables import WINDOWS_COLUMNS, read_tsv, write_tsv

MIN_MATE_MAPQ = 30
WINDOW_SIZE = 1000
WINDOW_STEP = 500


def load_discordant(path):
    if not path or path == "NULL" or not os.path.exists(path):
        return pd.DataFrame(
            columns=[
                "read_name",
                "mate_chr",
                "mate_position",
                "mate_mapq",
                "mate_strand",
            ]
        )
    df = read_tsv(path)
    for col in ["read_name", "mate_chr", "mate_position", "mate_mapq", "mate_strand"]:
        if col not in df.columns:
            df[col] = ""
    df["mate_mapq"] = pd.to_numeric(df["mate_mapq"], errors="coerce").fillna(0)
    df["mate_position"] = pd.to_numeric(df["mate_position"], errors="coerce")
    df = df[df["mate_mapq"] > MIN_MATE_MAPQ]
    df = df[df["mate_position"].notna()]
    return df


def build_windows(df):
    if df.empty:
        return pd.DataFrame(
            columns=["window", "chrom", "chromStart", "chromEnd", "strand"]
        )

    windows = []
    for (chrom, strand), group in df.groupby(["mate_chr", "mate_strand"], dropna=False):
        positions = group["mate_position"].astype(int).tolist()
        for pos in positions:
            start = (pos // WINDOW_STEP) * WINDOW_STEP
            for offset in [0, -WINDOW_STEP]:
                win_start = start + offset
                if win_start < 0:
                    continue
                win_end = win_start + WINDOW_SIZE
                if win_start <= pos < win_end:
                    windows.append(
                        {
                            "window": f"{chrom}_{win_start}_{strand}",
                            "chrom": chrom,
                            "chromStart": win_start,
                            "chromEnd": win_end,
                            "strand": strand,
                        }
                    )
    return pd.DataFrame(windows).drop_duplicates()


def count_windows(df, windows, name_column):
    if windows.empty:
        windows = windows.copy()
        windows["count"] = 0
        windows[name_column] = [set() for _ in range(len(windows))]
        return windows
    df = df.copy()
    df["mate_position"] = df["mate_position"].astype(int)

    windows = windows.copy()
    windows["count"] = 0
    windows[name_column] = [set() for _ in range(len(windows))]
    for idx, row in windows.iterrows():
        chrom = row["chrom"]
        strand = row["strand"]
        start = row["chromStart"]
        end = row["chromEnd"]
        mask = (
            (df["mate_chr"] == chrom)
            & (df["mate_strand"] == strand)
            & (df["mate_position"] >= start)
            & (df["mate_position"] < end)
        )
        read_names = set(df.loc[mask, "read_name"].astype(str).tolist())
        windows.at[idx, "count"] = int(len(read_names))
        windows.at[idx, name_column] = read_names
    return windows


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-t", "--discordantReadFileTumor", required=True)
    parser.add_argument("-c", "--discordantReadFileControl", required=True)
    parser.add_argument("-b", "--blacklist_file", required=True)
    parser.add_argument("-o", "--outFile", required=True)
    parser.add_argument("-f", "--function_file", required=False)
    args = parser.parse_args()

    df = compute_windows(
        args.discordantReadFileTumor,
        args.discordantReadFileControl,
        args.blacklist_file,
        args.outFile,
    )
    write_tsv(df, args.outFile, WINDOWS_COLUMNS)


def compute_windows(
    tumor_path: str,
    control_path: str,
    blacklist_path: str,
    output_path: str,
) -> pd.DataFrame:
    tumor_df = load_discordant(tumor_path)
    control_df = load_discordant(control_path)

    pid = os.path.basename(output_path).replace(
        "_discordant_reads_1_kb_windows.tsv", ""
    )

    windows = build_windows(pd.concat([tumor_df, control_df], ignore_index=True))
    tumor_counts = count_windows(tumor_df, windows, "_tumor_read_names").rename(
        columns={"count": "tumor_discordant_read_count"}
    )
    control_counts = count_windows(control_df, windows, "_control_read_names").rename(
        columns={"count": "control_discordant_read_count"}
    )

    merged_counts = windows.merge(
        tumor_counts[["window", "tumor_discordant_read_count", "_tumor_read_names"]],
        on="window",
        how="left",
    ).merge(
        control_counts[
            ["window", "control_discordant_read_count", "_control_read_names"]
        ],
        on="window",
        how="left",
    )
    merged_counts["tumor_discordant_read_count"] = (
        merged_counts["tumor_discordant_read_count"].fillna(0).astype(int)
    )
    merged_counts["control_discordant_read_count"] = (
        merged_counts["control_discordant_read_count"].fillna(0).astype(int)
    )
    merged_counts["_tumor_read_names"] = merged_counts["_tumor_read_names"].apply(
        lambda x: x if isinstance(x, set) else set()
    )
    merged_counts["_control_read_names"] = merged_counts["_control_read_names"].apply(
        lambda x: x if isinstance(x, set) else set()
    )
    merged_counts["PID"] = pid

    merged_counts["blacklisted"] = ""
    if blacklist_path and os.path.exists(blacklist_path):
        blacklist_df = read_tsv(blacklist_path)
        if "window" in blacklist_df.columns:
            blacklist = set(blacklist_df["window"].tolist())
            merged_counts["blacklisted"] = merged_counts["window"].apply(
                lambda w: "yes" if w in blacklist else "no"
            )

    if "_tumor_read_names" in merged_counts.columns:
        merged_counts = merged_counts.drop(columns=["_tumor_read_names"])
    if "_control_read_names" in merged_counts.columns:
        merged_counts = merged_counts.drop(columns=["_control_read_names"])
    if not merged_counts.empty:
        merged_counts = merged_counts.sort_values(["chrom", "chromStart", "strand"])
    return merged_counts


if __name__ == "__main__":
    main()

File: src/count_discordant_reads.R
# Author: Lina Sieverling
# usage: R --no-save --slave --args -t <discordantReadFileTumor> -c <discordantReadFileControl> -b <blacklist_file> -o <outFile> -f <function_file> < count_discordant_reads.R
# description: - uses the output tables of add_mate_mapq.py
#              - gets the number of discordant reads in 1 kb windows
#              - merges adjacent windows with discordant reads in the tumor sample
#              - annotates merged windows with ENCODE blacklist overlap fraction
#              - output: table for each pid containing raw discordant read counts for tumor and control

library("optparse")
library("data.table")

option_list <- list(
  make_option(c("-t", "--discordantReadFileTumor"), type = "character", default = NULL,
              help = "Discordant read file for tumor", metavar = "character"),
  make_option(c("-c", "--discordantReadFileControl"), type = "character", default = NULL,
              help = "Discordant read file for control", metavar = "character"),
  make_option(c("-b", "--blacklist_file"), type = "character", default = NULL,
              help = "Blacklist file (BED/BED.GZ or legacy window table)", metavar = "character"),
  make_option(c("-o", "--outFile"), type = "character", default = NULL,
              help = "Output file", metavar = "character"),
  make_option(c("-f", "--function_file"), type = "character",
              default = "/abi/data/sieverling/global_scripts/functions.R",
              help = "File with R functions [default= %default]", metavar = "character")
)

opt_parser <- OptionParser(option_list = option_list)
opt <- parse_args(opt_parser)

discordantReadFileTumor   <- opt$discordantReadFileTumor
discordantReadFileControl <- opt$discordantReadFileControl
blacklist_file            <- opt$blacklist_file
outFile                   <- opt$outFile
function_file             <- opt$function_file

source(function_file)
options(scipen = 999)

# ------------------------------------------------------------------------------------
# helpers for ENCODE blacklist (hg19-blacklist.v2.bed.gz)
# ------------------------------------------------------------------------------------

normalize_chrom <- function(x) {
  x <- as.character(x)
  x <- sub("^chr", "", x, ignore.case = TRUE)
  x
}

read_blacklist_as_bed <- function(path) {
  if (is.null(path) || is.na(path) || path == "" || path == "no_file" || !file.exists(path)) {
    return(NULL)
  }

  con <- if (grepl("\\.gz$", path, ignore.case = TRUE)) gzfile(path, open = "rt") else file(path, open = "rt")
  on.exit(close(con))

  bl <- tryCatch(
    read.table(con, header = FALSE, sep = "\t", stringsAsFactors = FALSE, comment.char = "", quote = ""),
    error = function(e) NULL
  )

  if (is.null(bl) || nrow(bl) == 0 || ncol(bl) < 3) return(NULL)

  # BED-like check: col2/col3 numeric
  s2 <- suppressWarnings(as.numeric(bl[[2]]))
  s3 <- suppressWarnings(as.numeric(bl[[3]]))
  if (all(is.na(s2)) || all(is.na(s3))) return(NULL)

  out <- data.frame(
    chrom = normalize_chrom(bl[[1]]),
    start = s2,
    end   = s3,
    stringsAsFactors = FALSE
  )
  out <- out[!is.na(out$start) & !is.na(out$end) & out$end > out$start, ]
  if (nrow(out) == 0) return(NULL)
  out
}

read_blacklist_as_window_table <- function(path) {
  # legacy format with column "window"
  if (is.null(path) || is.na(path) || path == "" || path == "no_file" || !file.exists(path)) {
    return(NULL)
  }
  x <- tryCatch(
    read.table(path, header = TRUE, sep = "\t", stringsAsFactors = FALSE, comment.char = ""),
    error = function(e) NULL
  )
  if (is.null(x) || !("window" %in% colnames(x))) return(NULL)
  x
}

compute_blacklist_fraction <- function(chrom, start, end, bl_by_chr) {
  width <- end - start
  if (is.na(width) || width <= 0) return(0)

  chr_key <- as.character(chrom)
  if (!(chr_key %in% names(bl_by_chr))) return(0)

  bl <- bl_by_chr[[chr_key]]
  if (is.null(bl) || nrow(bl) == 0) return(0)

  ov_start <- pmax(start, bl$start)
  ov_end <- pmin(end, bl$end)
  overlap <- pmax(0, ov_end - ov_start)
  covered <- sum(overlap)   # safe for Boyle v2 (non-overlapping intervals)

  covered / width
}

# ------------------------------------------------------------------------------------
# get 1 kb windows with discordant read count
# ------------------------------------------------------------------------------------

windowCountList <- list()
discordantReadFileList <- list(
  tumor   = discordantReadFileTumor,
  control = discordantReadFileControl
)

for (sample in c("tumor", "control")) {

  discordantReadFile <- discordantReadFileList[[sample]]

  # check if file exists, if not: skip sample
  if (!file.exists(discordantReadFile)) {
    windowCount <- data.frame(window=NA, stringsAsFactors=FALSE)
    windowCount[, paste0(sample, '_discordant_read_count')] <- NA
    windowCountList[[sample]] <- windowCount[numeric(0), ]
    next
  }

  discordantReads_all <- read.table(discordantReadFile, header=TRUE, sep="\t", comment.char='', stringsAsFactors=FALSE)

  # only keep those with mate mapping quality > 30
  discordantReads <- subset(discordantReads_all, mate_mapq > 30)

  if (nrow(discordantReads) == 0) {
    windowCount <- data.frame(window=character(0), stringsAsFactors=FALSE)
    windowCount[, paste0(sample, '_discordant_read_count')] <- numeric(0)
    windowCountList[[sample]] <- windowCount
    next
  }

  # get 1kb window
  discordantReads$mate_position_1kb <- floor(discordantReads$mate_position / 1000) * 1000
  discordantReads$window <- paste0(discordantReads$mate_chr, "_", as.character(discordantReads$mate_position_1kb), "_", discordantReads$mate_strand)

  # get read count per window
  windowCount <- as.data.frame(table(discordantReads$window), stringsAsFactors=FALSE)
  colnames(windowCount) <- c("window", paste0(sample, "_discordant_read_count"))
  windowCountList[[sample]] <- windowCount
}

# merge tumor/control window counts and set missing to 0
windowCountMerged <- merge(windowCountList$tumor, windowCountList$control, by="window", all=TRUE)
windowCountMerged[is.na(windowCountMerged)] <- 0

# edge case: no windows at all -> write valid empty output and exit success
if (nrow(windowCountMerged) == 0) {
  emptyTable <- data.frame(
    PID = character(0),
    window = character(0),
    chrom = character(0),
    chromStart = numeric(0),
    chromEnd = numeric(0),
    strand = character(0),
    tumor_discordant_read_count = numeric(0),
    control_discordant_read_count = numeric(0),
    blacklist_overlap_frac = numeric(0),
    blacklist_excluded = logical(0),
    blacklisted = character(0),
    stringsAsFactors = FALSE
  )
  write.table(emptyTable, file=outFile, quote=FALSE, row.names = FALSE, sep="\t")
  quit(save="no", status=0)
}

# add chromosome/start/end/strand
windowCountMerged$chrom <- gsub("_.*", "", windowCountMerged$window)
windowCountMerged$chromStart <- as.numeric(gsub("_", "", regmatches(windowCountMerged$window, regexpr("_.*_", windowCountMerged$window))))
windowCountMerged$chromEnd <- windowCountMerged$chromStart + 1000
windowCountMerged$strand <- gsub(".*_", "", windowCountMerged$window)

# add PID and order columns
windowCountMerged$PID <- gsub("_discordant_reads_1_kb_windows.tsv", "", basename(outFile))
windowCountMerged <- windowCountMerged[, c(
  "PID", "window", "chrom", "chromStart", "chromEnd", "strand",
  "tumor_discordant_read_count", "control_discordant_read_count"
)]
row.names(windowCountMerged) <- windowCountMerged$window

# ------------------------------------------------------------------------------------
# merge adjacent 1 kb windows with discordant reads in tumor sample
# (kept as in original logic, now before blacklist annotation)
# ------------------------------------------------------------------------------------

for (window1 in windowCountMerged[windowCountMerged$tumor_discordant_read_count != 0, "window"]) {

  # skip if row removed during prior merge
  if (!window1 %in% windowCountMerged$window) next

  while (1) {

    chromEnd <- windowCountMerged[window1, "chromEnd"]
    chrom <- windowCountMerged[window1, "chrom"]
    strand <- windowCountMerged[window1, "strand"]

    # adjacent downstream window on same chrom and strand
    window2 <- windowCountMerged[
      windowCountMerged$chrom == chrom &
        windowCountMerged$chromStart == chromEnd &
        windowCountMerged$strand == strand, "window"
    ]

    if (identical(window2, character(0))) break

    # require both windows to have non-zero tumor support
    if (windowCountMerged[window1, "tumor_discordant_read_count"] == 0 ||
        windowCountMerged[window2, "tumor_discordant_read_count"] == 0) break

    # merge counts + extend interval
    windowCountMerged[window1, "chromEnd"] <- windowCountMerged[window2, "chromEnd"]
    windowCountMerged[window1, "tumor_discordant_read_count"] <-
      windowCountMerged[window1, "tumor_discordant_read_count"] + windowCountMerged[window2, "tumor_discordant_read_count"]
    windowCountMerged[window1, "control_discordant_read_count"] <-
      windowCountMerged[window1, "control_discordant_read_count"] + windowCountMerged[window2, "control_discordant_read_count"]

    # remove second row
    windowCountMerged <- windowCountMerged[windowCountMerged$window != window2, ]
  }
}

# ------------------------------------------------------------------------------------
# annotate blacklist overlap
# ------------------------------------------------------------------------------------

# defaults
# (use rep() rather than a bare scalar: assigning a scalar to a new column on a
# 0-row data.frame errors in base R - "replacement has 1 row, data has 0")
windowCountMerged$blacklist_overlap_frac <- rep(0, nrow(windowCountMerged))
windowCountMerged$blacklist_excluded <- rep(FALSE, nrow(windowCountMerged))
windowCountMerged$blacklisted <- rep("no", nrow(windowCountMerged))  # legacy-compatible column

# try ENCODE BED(.gz) first
blacklist_bed <- read_blacklist_as_bed(blacklist_file)

if (!is.null(blacklist_bed) && nrow(blacklist_bed) > 0) {

  cand_chr_norm <- normalize_chrom(windowCountMerged$chrom)
  blacklist_bed <- blacklist_bed[blacklist_bed$chrom %in% unique(cand_chr_norm), , drop=FALSE]

  if (nrow(blacklist_bed) > 0) {
    bl_by_chr <- split(blacklist_bed, blacklist_bed$chrom)

    frac <- mapply(
      compute_blacklist_fraction,
      chrom = cand_chr_norm,
      start = windowCountMerged$chromStart,
      end   = windowCountMerged$chromEnd,
      MoreArgs = list(bl_by_chr = bl_by_chr)
    )

    windowCountMerged$blacklist_overlap_frac <- as.numeric(frac)
    windowCountMerged$blacklist_excluded <- windowCountMerged$blacklist_overlap_frac >= 0.5
    windowCountMerged$blacklisted <- ifelse(windowCountMerged$blacklist_excluded, "yes", "no")
  }

} else {
  # fallback: legacy blacklist table containing "window"
  legacy_blacklist <- read_blacklist_as_window_table(blacklist_file)
  if (!is.null(legacy_blacklist)) {
    windowCountMerged$blacklisted <- ifelse(windowCountMerged$window %in% legacy_blacklist$window, "yes", "no")
    windowCountMerged$blacklist_excluded <- windowCountMerged$blacklisted == "yes"
    windowCountMerged$blacklist_overlap_frac <- ifelse(windowCountMerged$blacklist_excluded, 1, 0)
  } else {
    # no usable blacklist provided
    windowCountMerged$blacklisted <- rep(NA, nrow(windowCountMerged))
    windowCountMerged$blacklist_excluded <- rep(FALSE, nrow(windowCountMerged))
    windowCountMerged$blacklist_overlap_frac <- rep(0, nrow(windowCountMerged))
  }
}

# ------------------------------------------------------------------------------------
# save results
# ------------------------------------------------------------------------------------
write.table(windowCountMerged, file=outFile, quote=FALSE, row.names = FALSE, sep="\t")